# Akıllı Telefon Bağımlılığı Tahmini - Veri Keşfi (EDA)

Hedef: `addicted_label` (0/1) sınıflandırması. Bu notebook'ta verinin yapısını, sütunlarını ve örnek değerlerini inceliyoruz.

In [1]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None)
pd.set_option('display.width', 250)
pd.set_option('display.max_rows', 60)

train = pd.read_csv('data/train.csv')
test  = pd.read_csv('data/test.csv')
sub   = pd.read_csv('data/sample_submission.csv')

print('train:', train.shape)
print('test :', test.shape)
print('sub  :', sub.shape)

train: (691369, 14)
test : (296302, 13)
sub  : (296302, 2)


## 1. Sütunlar ve Örnek Değerler

In [2]:
print('TÜM SÜTUNLAR:')
for c in train.columns:
    print('  -', c)

print('\nSÜTUN TİPLERİ:')
print(train.dtypes)

print('\nÖRNEK SATIRLAR (ilk 5):')
display(train.head(5))

TÜM SÜTUNLAR:
  - id
  - age
  - daily_screen_time_hours
  - social_media_hours
  - gaming_hours
  - work_study_hours
  - sleep_hours
  - notifications_per_day
  - app_opens_per_day
  - weekend_screen_time
  - gender
  - stress_level
  - academic_work_impact
  - addicted_label

SÜTUN TİPLERİ:
id                           int64
age                        float64
daily_screen_time_hours    float64
social_media_hours         float64
gaming_hours               float64
work_study_hours           float64
sleep_hours                float64
notifications_per_day      float64
app_opens_per_day          float64
weekend_screen_time        float64
gender                      object
stress_level                object
academic_work_impact        object
addicted_label               int64
dtype: object

ÖRNEK SATIRLAR (ilk 5):


,id,age,daily_screen_time_hours,social_media_hours,gaming_hours,work_study_hours,sleep_hours,notifications_per_day,app_opens_per_day,weekend_screen_time,gender,stress_level,academic_work_impact,addicted_label
0,0,24.0,NaN,1.83,1.59,2.11,7.46,122.0,38.0,8.63,Male,Medium,No,1
1,1,19.0,5.97,1.08,NaN,3.03,8.22,76.0,19.0,NaN,Female,Medium,No,0
2,2,18.0,5.09,NaN,NaN,NaN,6.25,134.0,60.0,7.47,Female,Low,Yes,0
3,3,21.0,6.42,1.26,1.42,3.36,8.85,112.0,94.0,8.66,Other,Low,NaN,1
4,4,26.0,11.20,1.87,2.81,1.95,5.25,NaN,NaN,13.39,Female,Medium,No,1


In [3]:
print('SON 5 SATIR:')
display(train.tail(5))

SON 5 SATIR:


,id,age,daily_screen_time_hours,social_media_hours,gaming_hours,work_study_hours,sleep_hours,notifications_per_day,app_opens_per_day,weekend_screen_time,gender,stress_level,academic_work_impact,addicted_label
691364,691364,22.0,10.78,3.25,1.57,1.65,7.71,86.0,94.0,12.89,Male,NaN,No,1
691365,691365,20.0,3.45,1.86,0.89,0.30,6.59,221.0,42.0,8.47,Female,High,No,1
691366,691366,NaN,NaN,NaN,NaN,2.06,6.16,NaN,86.0,7.79,Other,Medium,No,1
691367,691367,22.0,11.78,3.69,0.23,1.35,5.53,171.0,62.0,7.71,Other,NaN,NaN,1
691368,691368,21.0,5.64,1.03,3.05,1.01,4.91,169.0,NaN,9.47,Other,Medium,Yes,0


## 2. Hedef (addicted_label) Dağılımı

In [4]:
print(train['addicted_label'].value_counts())
print('Bağımlı oranı: {:.2%}'.format(train['addicted_label'].mean()))

addicted_label
1    490474
0    200895
Name: count, dtype: int64
Bağımlı oranı: 70.94%


## 3. Eksik (NaN) Değerler

In [5]:
na = train.isna().sum()
na_pct = (train.isna().mean() * 100).round(2)
na_df = pd.DataFrame({'NaN sayısı': na, 'NaN %': na_pct})
na_df = na_df[na_df['NaN sayısı'] > 0].sort_values('NaN sayısı', ascending=False)
print('Eksik değer içeren sütunlar (train):')
display(na_df)

print('\nTest setinde eksik değerler:')
display(pd.DataFrame({'NaN %': (test.isna().mean()*100).round(2)}).sort_values('NaN %', ascending=False).head(12))

Eksik değer içeren sütunlar (train):


,NaN sayısı,NaN %
social_media_hours,133995,19.38
gaming_hours,126821,18.34
weekend_screen_time,112063,16.21
daily_screen_time_hours,95854,13.86
app_opens_per_day,80710,11.67
notifications_per_day,67584,9.78
stress_level,55148,7.98
work_study_hours,51518,7.45
sleep_hours,44480,6.43
academic_work_impact,44224,6.40



Test setinde eksik değerler:


,NaN %
gaming_hours,20.05
weekend_screen_time,17.11
social_media_hours,16.00
notifications_per_day,11.55
daily_screen_time_hours,11.07
work_study_hours,9.37
app_opens_per_day,8.68
academic_work_impact,8.68
sleep_hours,7.58
stress_level,6.62


## 4. Sayısal Değişkenlerin Özeti (describe)

In [6]:
print('Sayısal sütunların özeti:')
display(train.describe().T)

Sayısal sütunların özeti:


,count,mean,std,min,25%,50%,75%,max
id,691369.0,345684.000000,199581.183467,0.00,172842.00,345684.00,518526.00,691368.00
age,662440.0,26.615408,5.153162,18.00,22.00,27.00,31.00,35.00
daily_screen_time_hours,595515.0,7.640865,2.721446,0.50,5.48,7.77,9.84,15.00
social_media_hours,557374.0,2.471038,1.316137,0.00,1.45,2.31,3.37,8.00
gaming_hours,564548.0,1.459265,0.934552,0.00,0.70,1.33,2.09,4.00
work_study_hours,639851.0,2.366971,1.258797,0.00,1.36,2.20,3.20,6.00
sleep_hours,646889.0,6.804334,1.234512,4.50,5.78,6.80,7.87,9.00
notifications_per_day,623785.0,145.894900,65.917556,20.00,93.00,150.00,204.00,250.00
app_opens_per_day,610659.0,102.636781,48.093970,15.00,64.00,104.00,145.00,180.00
weekend_screen_time,579306.0,9.479866,2.856006,0.51,7.28,9.58,11.75,17.56


In [7]:
print('Kategorik sütunların unique değerleri:')
for c in ['gender', 'stress_level', 'academic_work_impact']:
    vals = sorted(train[c].dropna().unique().tolist())
    print(f'  {c}: {vals}')

print('\nKategorik sütunların dağılımı:')
for c in ['gender', 'stress_level', 'academic_work_impact']:
    print(f'\n--- {c} ---')
    display(pd.crosstab(train[c], train['addicted_label'], margins=True))

Kategorik sütunların unique değerleri:
  gender: ['Female', 'Male', 'Other']
  stress_level: ['High', 'Low', 'Medium']
  academic_work_impact: ['No', 'Yes']

Kategorik sütunların dağılımı:

--- gender ---


addicted_label,0,1,All
gender,,,
Female,65628,155967,221595
Male,61910,161752,223662
Other,64902,152176,217078
All,192440,469895,662335



--- stress_level ---


addicted_label,0,1,All
stress_level,,,
High,63737,157136,220873
Low,60016,147767,207783
Medium,61094,146471,207565
All,184847,451374,636221



--- academic_work_impact ---


addicted_label,0,1,All
academic_work_impact,,,
No,91482,225097,316579
Yes,96564,234002,330566
All,188046,459099,647145


## 5. Bağımlı vs Bağımsız Grup Karşılaştırması (Sayısal)

In [8]:
num_cols = ['age', 'daily_screen_time_hours', 'social_media_hours', 'gaming_hours',
            'work_study_hours', 'sleep_hours', 'notifications_per_day',
            'app_opens_per_day', 'weekend_screen_time']

grp = train.groupby('addicted_label')[num_cols].mean().T
grp.columns = ['Bağımsız (0)', 'Bağımlı (1)']
grp['Fark (1-0)'] = grp['Bağımlı (1)'] - grp['Bağımsız (0)']
display(grp.round(2))

,Bağımsız (0),Bağımlı (1),Fark (1-0)
age,26.58,26.63,0.05
daily_screen_time_hours,5.04,8.71,3.66
social_media_hours,1.38,2.92,1.54
gaming_hours,1.16,1.58,0.42
work_study_hours,1.87,2.57,0.70
sleep_hours,6.72,6.84,0.12
notifications_per_day,147.09,145.41,-1.68
app_opens_per_day,97.87,104.59,6.72
weekend_screen_time,6.85,10.56,3.71


# Taban Model: LightGBM + 5-Fold CV

Metrik: **ROC-AUC**. LightGBM NaN değerleri yerleşik işler, kategorikleri `category` tipiyle alır.
Stratified 5-Fold CV ile OOF (out-of-fold) AUC hesaplayıp test tahmini üretiyoruz.

In [9]:
import numpy as np
import lightgbm as lgb
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import roc_auc_score

feats = [c for c in train.columns if c not in ['id', 'addicted_label']]
X = train[feats].copy()
X_test = test[feats].copy()

cat_cols = ['gender', 'stress_level', 'academic_work_impact']
for c in cat_cols:
    X[c] = X[c].astype('category')
    X_test[c] = X_test[c].astype('category')
    cats = X[c].cat.categories.union(X_test[c].cat.categories)
    X[c] = X[c].cat.set_categories(cats)
    X_test[c] = X_test[c].cat.set_categories(cats)

y = train['addicted_label'].values
print('X:', X.shape, '| X_test:', X_test.shape)

X: (691369, 12) | X_test: (296302, 12)


In [10]:
skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
oof = np.zeros(len(X))
test_pred = np.zeros(len(X_test))

params = dict(
    objective='binary',
    metric='auc',
    learning_rate=0.05,
    num_leaves=63,
    min_child_samples=100,
    subsample=0.9,
    colsample_bytree=0.9,
    n_estimators=3000,
    random_state=42,
    verbosity=-1,
)

for fold, (tr_idx, va_idx) in enumerate(skf.split(X, y)):
    X_tr, X_va = X.iloc[tr_idx], X.iloc[va_idx]
    y_tr, y_va = y[tr_idx], y[va_idx]
    model = lgb.LGBMClassifier(**params)
    model.fit(X_tr, y_tr, eval_set=[(X_va, y_va)], eval_metric='auc',
              callbacks=[lgb.early_stopping(100, verbose=False)])
    oof[va_idx] = model.predict_proba(X_va)[:, 1]
    test_pred += model.predict_proba(X_test)[:, 1] / skf.n_splits
    print(f'Fold {fold+1} AUC: {roc_auc_score(y_va, oof[va_idx]):.5f}')

print(f'\nCV OOF AUC: {roc_auc_score(y, oof):.5f}')
print(f'Best iter ort.: {np.mean([m.best_iteration_ for m in [model]])}')

Fold 1 AUC: 0.96262


Fold 2 AUC: 0.96337


Fold 3 AUC: 0.96379


Fold 4 AUC: 0.96437


Fold 5 AUC: 0.96330

CV OOF AUC: 0.96349
Best iter ort.: 1556.0


In [11]:
imp = pd.Series(model.feature_importances_, index=feats).sort_values(ascending=False)
display(imp.to_frame('importance'))

,importance
notifications_per_day,13582
app_opens_per_day,12474
weekend_screen_time,11161
daily_screen_time_hours,11050
work_study_hours,10569
social_media_hours,10165
gaming_hours,9503
sleep_hours,9303
age,4950
gender,1346


In [12]:
sub = pd.DataFrame({'id': test['id'], 'addicted_label': test_pred})
sub.to_csv('submission_lgbm_baseline.csv', index=False)
print(sub.shape)
print(sub.head())
print(sub['addicted_label'].describe())

(296302, 2)
       id  addicted_label
0  691369        0.999373
1  691370        0.966710
2  691371        0.954159
3  691372        0.989621
4  691373        0.997836
count    296302.000000
mean          0.709957
std           0.366307
min           0.000097
25%           0.394729
50%           0.943135
75%           0.999541
max           1.000000
Name: addicted_label, dtype: float64


# Eksik (NaN) Değer Yönetimi Stratejisi

Her sütun için boşluk oranı ve diğer sütunlar kullanılarak ne kadar iyi tahmin edilebildiği (R²) ölçüldü. R², o sütundaki boş değerlerin diğer 11 sütundan öğrenilen bir LightGBM regresyonu ile kurtarılabileceğini gösterir.

| Sütun | NaN% (train/test) | Tahmin edilebilirlik R² | Strateji |
|---|---|---|---|
| `daily_screen_time_hours` | 13.9 / 11.1 | **0.81** | Diğer sütunlardan impute edilir (çok güvenilir) + NaN bayrağı |
| `weekend_screen_time` | 16.2 / 17.1 | 0.63 | LightGBM impute (iyi) + NaN bayrağı |
| `social_media_hours` | 19.4 / 16.0 | 0.54 | LightGBM impute (kısmen) + NaN bayrağı |
| `work_study_hours` | 7.5 / 9.4 | 0.44 | LightGBM impute (kısmen) + NaN bayrağı |
| `gaming_hours` | 18.3 / 20.1 | 0.41 | LightGBM impute (kısmen) + NaN bayrağı |
| `sleep_hours` | 6.4 / 7.6 | **0.02** | İmpute ETME — LightGBM NaN'ı native işler, sadece NaN bayrağı |
| `notifications_per_day` | 9.8 / 11.6 | **0.02** | İmpute ETME — sadece NaN bayrağı |
| `app_opens_per_day` | 11.7 / 8.7 | **0.02** | İmpute ETME — sadece NaN bayrağı |
| `age` | 4.2 / 5.8 | **0.00** | İmpute ETME — sadece NaN bayrağı |
| `stress_level` | 8.0 / 6.6 | - (kategorik) | Mode impute + NaN bayrağı (veya native) |
| `academic_work_impact` | 6.4 / 8.7 | - (kategorik) | Mode impute + NaN bayrağı (veya native) |
| `gender` | 4.2 / 4.8 | - (kategorik) | Mode impute + NaN bayrağı (veya native) |

**Anahtar bulgu:** Eksiklik bilgisi kendisi tahmin gücü taşır. Hangi sütunun boş olduğu (`is_missing` bayrakları) modele ayrı özellik olarak eklenmeli. Düşük R²'li sütunlarda imputasyon gürültü ekler; LightGBM NaN'ı doğrudan işlediği için o sütunları ham haliyle bırakmak en güvenlisi.

In [13]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from scipy.stats import pearsonr
import lightgbm as lgb

imp_test = train.copy()
for c in cat_cols:
    imp_test[c] = imp_test[c].astype('category')

targets = ['daily_screen_time_hours', 'weekend_screen_time', 'social_media_hours',
           'work_study_hours', 'gaming_hours', 'sleep_hours',
           'notifications_per_day', 'app_opens_per_day', 'age']
res = {}
for t in targets:
    others = [c for c in imp_test.columns if c not in ['id', 'addicted_label', t]]
    sub = imp_test[others + [t]].dropna(subset=[t]).sample(200000, random_state=42)
    y = sub[t].values
    X = sub[others]
    X_tr, X_va, y_tr, y_va = train_test_split(X, y, test_size=0.2, random_state=42)
    m = lgb.LGBMRegressor(n_estimators=300, learning_rate=0.05, num_leaves=31, verbosity=-1, random_state=42)
    m.fit(X_tr, y_tr)
    p = m.predict(X_va)
    res[t] = 1 - mean_squared_error(y_va, p) / np.var(y_va)

imp_df = pd.DataFrame(res.items(), columns=['Sütun', 'Impute R²']).sort_values('Impute R²', ascending=False)
display(imp_df.round(3))

,Sütun,Impute R²
0,daily_screen_time_hours,0.808
1,weekend_screen_time,0.629
2,social_media_hours,0.544
3,work_study_hours,0.439
4,gaming_hours,0.413
5,sleep_hours,0.022
6,notifications_per_day,0.017
7,app_opens_per_day,0.015
8,age,0.004
